### PR TITLE
Token analyzer cronjob

### PR DESCRIPTION
**Pull request by kayibal**
Analyzes all tokens that have not been analyzed periodically.

Currently the cronjob will:
- Fetch all tokens from the db in batches
- Distribute these batches to a configurable number of async threads
- Skip any token that has quality <= 5 (previously already failed to analyse)
- Skip any token that has gas set (previously already successfully analysed)
- Analyze the remaining tokens
- Update the db with the analysis results.

### CODE REVIEW

File: token-analyzer/src/trace_call.rs [hunk 1]
```diff
@@ -367,6 +385,85 @@ impl TraceCallDetector {
 
         Ok((TokenQuality::Good, Some(gas_per_transfer), Some(fees)))
     }
+
+    fn calculate_fee(
```
Comment (tvinagre) [inline token-analyzer/src/trace_call.rs:389]: Would be great to later add docs here with an overview of how fee is calculated

File: tycho-core/src/models/token.rs [hunk 2]
```diff
@@ -12,7 +12,9 @@ pub struct CurrencyToken {
     ///  - 100: Normal token
     ///  - 75: Rebase token
     ///  - 50: Fee token
-    ///  - 0: Scam token that we shouldn't use
+    ///  - 10: Token analysis failed at creation
+    ///  - 5: Token analysis failed on cronjob (after creation).
```
Comment (tvinagre) [inline tycho-core/src/models/token.rs:16]: Cronjob failure should be higher quality no?

Comment (tvinagre) [reply tycho-core/src/models/token.rs:16]: Doesn't matter anywayy

File: tycho-indexer/src/cli.rs [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+}
+
+#[derive(Args, Debug, Clone, PartialEq, Eq)]
+pub struct AnalyzeTokenArgs {
+    /// Blockchain to execute analysis for.
+    #[clap(long)]
+    pub chain: Chain,
+    /// How many concurrent threads to use for token analysis.
+    #[clap(long)]
+    pub concurrency: usize,
```
Comment (tvinagre) [inline tycho-indexer/src/cli.rs:119]: let him cook

File: tycho-indexer/src/cli.rs [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+        env,
+        hide_env_values = true,
+        default_value = "postgres://postgres:mypassword@localhost:5432/tycho_indexer_0"
+    )]
+    pub database_url: String,
+
+    /// Substreams API endpoint
+    #[clap(name = "endpoint", long, default_value = "https://mainnet.eth.streamingfast.io")]
+    pub endpoint_url: String,
+}
```
Comment (zizou0x) [inline tycho-indexer/src/cli.rs:60]: I think the substreams related thing should be in substreams subcommands structs as we don't need them for the token cronjob. Not mandatory to do it now though

File: tycho-indexer/src/extractor/evm/token_analysis_cron.rs [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+    rpc_url: &str,
+    gw: Arc<dyn ProtocolGateway + Send + Sync>,
+) -> anyhow::Result<()> {
+    let mut tokens = Vec::new();
+    let mut page = 0;
+    let page_size = analyze_args.fetch_batch_size as i64;
+    loop {
+        let start = Instant::now();
+        let pagination_params = PaginationParams::new(page, page_size);
+        tokens.clone_from(
```
Comment (zizou0x) [inline tycho-indexer/src/extractor/evm/token_analysis_cron.rs:25]: I didn't know this one, it's cool

File: tycho-indexer/src/extractor/evm/token_analysis_cron.rs [hunk 6]
```diff
@@ -0,0 +1,217 @@
...
+            &(gw.get_tokens(analyze_args.chain, None, None, Some(&pagination_params))
```
Comment (tvinagre) [inline tycho-indexer/src/extractor/evm/token_analysis_cron.rs:26]: shouldn't this be part of the parallelization?  Are we only parallelising the analysis?

File: tycho-indexer/src/extractor/evm/token_analysis_cron.rs [hunk 7]
```diff
@@ -0,0 +1,217 @@
...
+
+    let component = gw
+        .get_protocol_components_by_tokens(&chain, Some(&addresses), Some(100_000f64))
+        .await?;
+
+    let liquidity_token_owners = component
+        .into_iter()
+        .filter_map(|pc| {
+            let pool = if !pc.contract_addresses.is_empty() {
+                Some(H160::from_slice(&pc.contract_addresses[0]))
```
Comment (tvinagre) [inline tycho-indexer/src/extractor/evm/token_analysis_cron.rs:78]: Could this be wrong in newer implementations?

File: tycho-indexer/src/cli.rs (hunk 4 above)
Comment (kayibal) [reply tycho-indexer/src/cli.rs:60]: But this in cli module, not in a token cronjob specific module 😅

File: tycho-indexer/src/extractor/evm/token_analysis_cron.rs (hunk 5 above)
Comment (kayibal) [reply tycho-indexer/src/extractor/evm/token_analysis_cron.rs:25]: New clippy version thought me ;)

File: tycho-indexer/src/extractor/evm/token_analysis_cron.rs [hunk 8]
```diff
@@ -0,0 +1,217 @@
...
+            } else {
+                H160::from_str(&pc.id).ok()
+            };
+
+            if let Some(pool) = pool {
```
Comment (tvinagre) [inline tycho-indexer/src/extractor/evm/token_analysis_cron.rs:83]:
```rust
pool.map(|pool| {
    pc.tokens.iter().map(|t| {
        (H160::from_slice(t), (pool, U256::from(100_000)))
    })
})
```

File: tycho-indexer/src/extractor/evm/token_analysis_cron.rs [hunk 9]
```diff
@@ -0,0 +1,217 @@
...
+                None
+            }
+        })
+        .flatten()
+        .collect::<HashMap<_, _>>();
+    let tf = Arc::new(TokenFinder::new(liquidity_token_owners));
+    let analyzer = TraceCallDetector::new(eth_rpc_url.as_str(), tf);
+    for t in tokens.iter_mut() {
+        // Skip tokens that failed previously and ones we already analyzed successfully.
+        if t.quality <= 5 || !t.gas.is_empty() {
```
Comment (tvinagre) [inline tycho-indexer/src/extractor/evm/token_analysis_cron.rs:100]: This way we'll analyze every token that's already been set as 100, no? Can we optimize this?

File: tycho-indexer/src/extractor/evm/token_analysis_cron.rs (hunk 6 above)
Comment (kayibal) [reply tycho-indexer/src/extractor/evm/token_analysis_cron.rs:26]: I think it is not necessary to parallelize

File: tycho-indexer/src/extractor/evm/token_analysis_cron.rs (hunk 7 above)
Comment (kayibal) [reply tycho-indexer/src/extractor/evm/token_analysis_cron.rs:78]: Yes and this is a workaround, for newer implementations we should add link related contracts explicitly via `protocol_component_holds_contracts`.

**Pull request merged**